src/auth/firebase.service.ts

```diff
@@ -0,0 +1,156 @@
+import * as fs from 'fs';
+import { Injectable, Inject } from '@nestjs/common';
+import { initializeApp, App, cert } from 'firebase-admin/app';
+import { getAuth } from 'firebase-admin/auth';
+import { Firestore, getFirestore, Timestamp } from 'firebase-admin/firestore';
+
+import { FirebaseOptions } from './interfaces';
+import { MODULE_OPTIONS_TOKEN } from './firebase.module-definition';
+
+@Injectable()
+export class FirebaseService {
+  app: App;
+  db: Firestore;
+
+  constructor(
+    @Inject(MODULE_OPTIONS_TOKEN)
+    options?: FirebaseOptions,
+  ) {
+    const appOptions = options
+      ? {
+          credential: cert(
+            JSON.parse(fs.readFileSync(options.credentialFilePath, 'utf8')),
+          ),
+          databaseURL: options.databaseUri,
+        }
+      : {};
+
+    this.app = initializeApp(appOptions);
+    this.db = getFirestore(this.app);
+  }
+
+  async verifyIdToken(idToken: string) {
+    return getAuth(this.app).verifyIdToken(idToken);
+  }
+
+  async getUserInfo(uid: string) {
+    return this.db.collection('users').doc(uid).get();
+  }
+
+  private upsertElementByArticleId = (
+    articleId: string,
+    list: any[],
+    element: any,
+  ) => {
+    const newArray = [...list];
+    const index = list.findIndex((s) => s.articleId === articleId);
+    index !== -1 ? (newArray[index] = element) : newArray.push(element);
+    return newArray;
+  };
+
+  async updateArticleRecord(uid: string, articleId: string) {
+    const userRef = this.db.collection('articles').doc(uid);
+    const doc = await userRef.get();
+
+    const editedList = doc.get('edited') ?? [];
+    const subscribedList = doc.get('subscribed') ?? [];
+    const element = {
+      articleId,
+      timeStamp: Timestamp.now(),
+    };
+
+    const newEditedList = this.upsertElementByArticleId(
+      articleId,
+      editedList,
+      element,
+    );
+    const newSubscribedList = this.upsertElementByArticleId(
+      articleId,
+      subscribedList,
+      element,
+    );
+
+    userRef.set(
+      { edited: newEditedList, subscribed: newSubscribedList },
+      { merge: true },
+    );
+
+    return Promise.resolve({
+      edited: newEditedList,
+      subscribed: newSubscribedList,
+    });
+  }
+
+  async updateArticleSubscription(
+    uid: string,
+    articleId: string,
+    isSubscribe: boolean,
+  ) {
+    const userRef = this.db.collection('articles').doc(uid);
+    const doc = await userRef.get();
+
+    const subscribedList = doc.get('subscribed') ?? [];
```

Comment:
Could we write in some readable without “??”

Reply:
Do you mean explicitly check whether it os unknown by `if` statement?

Reply:
yup

Reply:
In this case, `doc.get('subscribed')` returns `undefined` if no such field exists.
So, I think using nullish coalescing operator is more concise and more readable.